destination/writer/writer.go

```diff
@@ -0,0 +1,248 @@
+// Copyright © 2022 Meroxa, Inc. & Yalantis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/conduitio-labs/conduit-connector-clickhouse/columntypes"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jmoiron/sqlx"
+)
+
+const (
+	// metadata key with the name of the table.
+	metadataTable = "clickhouse.table"
+
+	// query patterns.
+	insertFmt = "INSERT INTO %s (%s) VALUES (%s)"
+	updateFmt = "ALTER TABLE %s UPDATE %s WHERE %s"
+	deleteFmt = "ALTER TABLE %s DELETE WHERE %s"
+
+	coma        = ","
+	placeholder = "?"
+	operatorAnd = " AND "
+)
+
+// Writer implements a writer logic for ClickHouse destination.
+type Writer struct {
+	db          *sqlx.DB
+	table       string
+	columnTypes map[string]string
+}
+
+// Params is an incoming params for the New function.
+type Params struct {
+	DB    *sqlx.DB
+	Table string
+}
+
+// NewWriter creates new instance of the Writer.
+func NewWriter(ctx context.Context, params Params) (*Writer, error) {
+	writer := &Writer{
+		db:    params.DB,
+		table: params.Table,
+	}
+
+	columnTypes, err := columntypes.GetColumnTypes(ctx, writer.db, writer.table)
+	if err != nil {
+		return nil, fmt.Errorf("get column types: %w", err)
+	}
+
+	writer.columnTypes = columnTypes
+
+	return writer, nil
+}
+
+// Insert inserts a record.
+func (w *Writer) Insert(ctx context.Context, record sdk.Record) error {
+	tableName := w.getTableName(record.Metadata)
+
+	payload, err := w.structurizeData(record.Payload.After)
+	if err != nil {
+		return fmt.Errorf("structurize payload: %w", err)
+	}
+
+	// if payload is empty return empty payload error
+	if payload == nil {
+		return errNoPayload
+	}
+
+	payload, err = columntypes.ConvertStructureData(w.columnTypes, payload)
+	if err != nil {
+		return fmt.Errorf("convert structure data: %w", err)
+	}
+
+	columns, values := w.extractColumnsAndValues(payload)
+
+	placeholders := make([]string, len(columns))
+	for i := range columns {
+		placeholders[i] = placeholder
+	}
+
+	query := fmt.Sprintf(insertFmt, tableName, strings.Join(columns, coma), strings.Join(placeholders, coma))
```

Comment:
WDYT about using an SQL builder? It may make it safer, since it should cover more cases we may have not thought of.

Reply:
TBH, I didn't find an SQL builder for the ClickHouse.
CH has a different syntax than MySQL (`update` and `delete` operations made by the `alter table` statement). 
https://clickhouse.com/docs/en/sql-reference/statements/alter/update/

Reply:
It would be nice to add that as a comment in the code.

Reply:
Ok, I've added it [here](https://github.com/conduitio-labs/conduit-connector-clickhouse/commit/6fe5cb8ad447527d379ea01a27b361f6e5787ddc#diff-56ce8e35dba85c6a25097b48d543beb48df432ab546eb8a1a890c21ca92054afR32).